test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp

```diff
@@ -0,0 +1,517 @@
+        "GetFieldModifiers failed");
+    return (access_flags & ACC_STATIC) != 0;
+}
+
+static void verify_int_field(JNIEnv* env, jclass klass, jfieldID fid) {
+    char* name = nullptr;
+    char* sig = nullptr;
+    check_jvmti_status(env,
+        jvmti->GetFieldName(klass, fid, &name, &sig, nullptr),
+        "GetFieldName failed");
+
+    if (strcmp(sig, "I") != 0) {
+        printf("ERROR: field '%s' is not int ('%s')\n", name, sig);
+        fflush(0);
+        fatal(env, "unexpected field type");
+    }
+
+    deallocate(env, name);
+    deallocate(env, sig);
+}
+
+
+/*
+Per jvmtiHeapReferenceInfoField spec:
+If the referrer object is not an interface, then the field indices are determined as follows:
+    make a list of all the fields in C and its superclasses,
+    starting with all the fields in java.lang.Object and ending with all the fields in C.
+    Within this list, put the fields for a given class in the order returned by GetClassFields.
+    Assign the fields in this list indices n, n+1, ..., in order,
+    where n is the count of the fields in all the interfaces implemented by C.
+    Note that C implements all interfaces directly implemented by its superclasses;
+    as well as all superinterfaces of these interfaces.
+If the referrer object is an interface, then the field indices are determined as follows:
+    make a list of the fields directly declared in I.
+    Within this list, put the fields in the order returned by GetClassFields.
+    Assign the fields in this list indices n, n+1, ..., in order,
+    where n is the count of the fields in all the superinterfaces of I.
+
+'Klass' struct contains all required data to calculate field indeces.
+Also contains static field values.
+For each test class 'Klass' struct is created and pointer to it set as a tag for jclass object.
+*/
+
+struct Klass {
+    jclass klass;
+    char* name;
+    Klass* super_klass;
+
+    struct Field {
+        jfieldID id;
+        char* name;
+
+        // Field value for static fields (0 for instance fields).
+        // All fields in the test classes are 'int'.
+        jint value;
+
+        void init(JNIEnv* env, jclass klass, jfieldID fid);
+    };
+
+    // Fields of the class and its superclasses
+    // as described in jvmtiHeapReferenceInfoField spec.
+    Field* fields;
+    jint field_count;
+
+    // Interfaces implemented by this klass, superclasses and superinterfaces.
+    Klass** interfaces;
+    jint interface_count;
+
+    // Number of fields in all implemented interfaces.
+    jint interface_field_count;
+
+    static Klass* explore(JNIEnv* env, jclass klass);
+
+private:
+    // Initializes fields, field_count.
+    void explore_fields(JNIEnv* env);
+    // Initializes interfaces, interface_count.
+    void explore_interfaces(JNIEnv* env);
+
+    void print() const;
+};
+
+/*
+For each test object 'Object' struct is created and pointer to it set as a tag for jobject.
```

Comment:
```suggestion
For each test object, the 'Object' struct is created and a pointer to it is set as the jobject's tag.
```

Reply:
Fixed.
Updated comment for 'Klass' class the same way